lib/membrane_http_adaptive_stream/hls_source.ex

```diff
@@ -117,22 +125,151 @@ defmodule Membrane.HLS.Source do
     {[], %{state | client_genserver: client_genserver}}
   end
 
+  @impl true
+  def handle_pad_added(Pad.ref(pad_name, _id) = pad_ref, ctx, state)
+      when ctx.playback == :stopped do
+    state = state |> put_in([pad_name, :ref], pad_ref)
+    {[], state}
+  end
+
+  # todo: refactor error messages in handle_pad_added/3
+
+  @impl true
+  def handle_pad_added(Pad.ref(pad_name, _id) = pad_ref, ctx, state)
+      when ctx.playback == :playing do
+    case state.status do
+      :waiting_on_pads -> :ok
+      :running -> raise "dupa"
+    end
+
+    state = state |> put_in([pad_name, :ref], pad_ref)
+
+    if map_size(ctx.pads) == length(state.new_tracks_notification),
+      do: state |> start_running(),
+      else: {[], state}
+  end
+
   @impl true
   def handle_playing(_ctx, state) do
-    {[audio_stream_format], [video_stream_format]} =
+    if state.audio_output.ref != nil or state.video_output.ref != nil do
+      state |> start_running()
+    else
+      state |> generate_new_tracks_notification()
+    end
```

Comment:
It think that in scenario described by the following steps:
1. `handle_pad_added(:video, ...)` is invoked
2. `handle_playing()` is invoked
3. `handle_pad_added(:audio, ...)`

we will raise an error and at the same time I don't think we should prevent user from linking `:audio` pad when the element is already playing.